### PR TITLE
Generator of tensor inputs with variable layout and structure (batch/non-batch, hybrid/non-hybrid, block/non-block)

### PR DESCRIPTION
**Pull request by pearu**
This PR introduces `TestCase.generate_simple_inputs` method that is an improved and generalized version of the `TestSparseCompressed._generate_small_inputs` method.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89502
* #88733
* __->__ #88914

cc @nikitaved @cpuhrsch @amjames @bhosmer

### CODE REVIEW

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+                                                            enable_batched=enable_batched, enable_hybrid=enable_hybrid,
+                                                            enable_zero_sized=enable_zero_sized,
+                                                            enable_batched_variable_nse=enable_batched_variable_nse,
+                                                            output_tensor=False):
+                if layout is torch.strided:
+                    size = kwargs.pop('size', args[0].shape)  # to ensure that a zero-sized tensor has the desired shape
+                    assert len(args) == 1
+                    yield args[0].reshape(size)
+                elif layout is torch.sparse_coo:
+                    yield torch.sparse_coo_tensor(*args, **kwargs).coalesce()
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2491]: Testing uncoalesced tensors is of value.

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+            Expand values in base to blocks of values with densesize.
+            """
+            if not densesize:
+                return base
+            if isinstance(base, list):
+                return [expand_values(b, densesize) for b in base]
+
+            r = 0
+            for s in reversed(densesize):
+                r = [copy.deepcopy(r) for _ in range(s)]
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2510]: Hmm, why can't we just create `r` as `torch.zeros(densesize)`?

File: torch/testing/_internal/common_utils.py [hunk 3]
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+                    kwargs.update(layout=layout)
+                    yield torch.sparse_compressed_tensor(*args, **kwargs)
+                else:
+                    assert 0  # unreachable
+            return
+
+        def expand_values(base, densesize):
+            """
+            Expand values in base to blocks of values with densesize.
+            """
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2502]: I am confused. `base` seems to be a scalar.

File: torch/testing/_internal/common_utils.py [hunk 4]
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+    def generate_simple_inputs(self, layout,
+                               device=None,
+                               dtype=None,
+                               index_dtype=None,
+                               enable_batched=True,
+                               enable_hybrid=True,
+                               enable_zero_sized=True,
+                               enable_batched_variable_nse=False,
+                               output_tensor=True,
+                               patterns=None):
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2452]: What do you think about having `low/high` parameters for values?

File: torch/testing/_internal/common_utils.py [hunk 5]
```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+            if base == 0:
+                return r
+            for index in itertools.product(*map(range, densesize)):
+                v = base + sum(10**c * i for c, i in enumerate(index))
+                r1 = r
+                for i in range(len(densesize) - 1):
+                    r1 = r1[index[i]]
+                assert r1[index[-1]] == 0, r1[index[-1]]
+                r1[index[-1]] = v
+            return r
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/testing/_internal/common_utils.py:2520]: Maybe something like that is a bit easier to understand? Along with the comment about what kind of mapping is employed...
```python
In [3]: l = expand_values(1, (3, 4, 5))
                                                    
In [4]: torch.tensor(l)                             
Out[4]:                                
tensor([[[  1, 101, 201, 301, 401],                 
         [ 11, 111, 211, 311, 411],   
         [ 21, 121, 221, 321, 421],                                                                      
         [ 31, 131, 231, 331, 431]],                
                                                    
        [[  2, 102, 202, 302, 402],                 
         [ 12, 112, 212, 312, 412],          
         [ 22, 122, 222, 322, 422],                 
         [ 32, 132, 232, 332, 432]],         
                                                    
        [[  3, 103, 203, 303, 403],                                                                                                     
         [ 13, 113, 213, 313, 413], 
         [ 23, 123, 223, 323, 423],                 
         [ 33, 133, 233, 333, 433]]])
                                                    
In [5]: densesize = (3, 4, 5)       
                                                    
In [6]: x = torch.tensor(range(densesize[0]))       
                                                    
In [7]: for i, d in enumerate(densesize[1:]):       
   ...:     y = torch.tensor(range(d))       
   ...:     x = x[..., None] + (10 ** (i + 1)) * y[None, ...]                                                                                                 
   ...:                                      
                                                    
In [8]: x + 1                                                                                            
Out[8]:                                                             
tensor([[[  1, 101, 201, 301, 401],                                 
         [ 11, 111, 211, 311, 411],                                 
         [ 21, 121, 221, 321, 421],                                 
         [ 31, 131, 231, 331, 431]],                                           
                                                                    
        [[  2, 102, 202, 302, 402],                                            
         [ 12, 112, 212, 312, 412],                                            
         [ 22, 122, 222, 322, 422],                                            
         [ 32, 132, 232, 332, 432]],                                           

        [[  3, 103, 203, 303, 403],                                            
         [ 13, 113, 213, 313, 413],                                            
         [ 23, 123, 223, 323, 423],                                            
         [ 33, 133, 233, 333, 433]]])              
```

File: torch/testing/_internal/common_utils.py (hunk 4 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2452]: The aim is to have inputs where all values are distinct (to ease debugging and the development of sparse algorithms). I think `low/high` parameters make sense for randomly generated samples but here we want the inputs to be highly predictable.

File: torch/testing/_internal/common_utils.py (hunk 2 above)
Comment (nikitaved) [reply torch/testing/_internal/common_utils.py:2510]: It does not work with tensors directly? Also, the Tensor class has a `tolist` method that preserves it's nesteness.

Comment (nikitaved) [reply torch/testing/_internal/common_utils.py:2510]: I am not sure about that, now after looking at the test... And I am not sure we ever construct sparse tensors from something which is not TensorLike in ATen...

File: torch/testing/_internal/common_utils.py (hunk 4 above)
Comment (nikitaved) [reply torch/testing/_internal/common_utils.py:2452]: Knowing the range of values could be useful to control precision. The values generated seem to be only unique when each size `<10`, which is the case I assume, right?

File: torch/testing/_internal/common_utils.py (hunk 2 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2510]: You can construct tensors from lists, for instance:
```python
>>> torch.sparse_csr_tensor([0, 2, 4], [0, 1, 0, 1], [1, 2, 3, 4])
tensor(crow_indices=tensor([0, 2, 4]),
       col_indices=tensor([0, 1, 0, 1]),
       values=tensor([1, 2, 3, 4]), size=(2, 2), nnz=4,
       layout=torch.sparse_csr)
```
In tests and examples, we often construct tensors from Python lists, so I don't see the relevance of constructing these within ATen.

Comment (nikitaved) [reply torch/testing/_internal/common_utils.py:2510]: If using tensors leads to cleaner code, and when there is `tolist`, then why not?

File: torch/testing/_internal/common_utils.py (hunk 4 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2461]: yes, the generated values aim at being unique. Notice the sizes of the generated tensors are intentionally small so that a human could easily track how the tensor elements transform when applying sparse tensor conversion/indexing and other tools that act mostly on the tensor structure rather than on the values.

For tests where precision parameters matter, one should use the `make_tensor` or other tensor generators that provide the low/high parameter feature.

File: torch/testing/_internal/common_utils.py (hunk 1 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2500]: `coalesce()` method call is now removed.

File: torch/testing/_internal/common_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2529]: Now using:
```python
        def expand_values(base, densesize):
            """
            Expand values in base to blocks of values with densesize.
            """
            if not densesize:
                return base
            if isinstance(base, list):
                return torch.stack([expand_values(b, densesize) for b in base])
            if base == 0:
                return torch.zeros(densesize, dtype=torch.int64)
            r = torch.tensor(range(densesize[0]))
            for i, d in enumerate(densesize[1:]):
                y = torch.arange(d, dtype=torch.int64) * (10 ** (i + 1))
                r = r[..., None] + y[None, ...]
            r.add_(base)
            return r
```

File: torch/testing/_internal/common_utils.py (hunk 2 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2519]: > What about we create r as torch.zeros(densesize)? It appears that the output of this function is fed into torch.tensor anyway...

Done.

File: torch/testing/_internal/common_utils.py [hunk 6]
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+                            target[1].extend(d[1])
+                        else:
+                            for j in range(len(d)):
+                                target[j].append(d[j])
+            return batch_data
+
+        def expand_values(base, densesize):
+            """
+            Expand values in base to blocks of values with densesize.
+            """
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2609]:
```suggestion
        def generate_values(offset, densesize):
            """
            Generates a tensor of shape densesize with values
            at indices i_1, ..., i_d with i_j < densesize[j] for any 0 < j < len(densesize)
            equal to i_1 * 10^0 + ... + i_d * 10^{d - 1}.
            This mapping produces UNIQUE values as long as densesize[i] < 10.
            """
```
?

File: torch/testing/_internal/common_utils.py [hunk 7]
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+
+        def get_shape(pattern):
+            if isinstance(pattern, list):
+                if not pattern:
+                    return (0,)
+                return (len(pattern), *get_shape(pattern[0]))
+            return ()
+
+        def get_blockpattern(pattern, blocksize):
+            basesize = get_shape(pattern)
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2515]: nit: if pattern is a tensor, then there is no need in `get_shape`.

File: torch/testing/_internal/common_utils.py [hunk 8]
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+            assert basesize[0] % blocksize[0] == 0, (basesize, blocksize)
+            assert basesize[1] % blocksize[1] == 0, (basesize, blocksize)
+            blockpattern = [[0 for _ in range(basesize[1] // blocksize[1])] for _ in range(basesize[0] // blocksize[0])]
+            for i, row in enumerate(pattern):
+                for j, maskin in enumerate(row):
+                    bi = i // blocksize[0]
+                    bj = j // blocksize[1]
+                    blockno = 1 + bi * (basesize[1] // blocksize[1]) + bj
+                    blockpattern[bi][bj] = blockno if maskin else blockpattern[bi][bj]
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2524]: There could be several `i, j` that are mapped to the same `bi, bj`, which means that only the very last "valid" value will be written. It is not a concern here, right? Also nit: seems like this could be simplified with tensors and broadcasting.

File: torch/testing/_internal/common_utils.py [hunk 9]
```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+                for j, maskin in enumerate(row):
+                    if maskin:
+                        count += 1
+                        indices.append([i, j])
+                        values.append(count)
+                        strided_values[i][j] = count
+                        for i1 in range(i, basesize[0]):
+                            crow_indices[i1 + 1] += 1
+                        for j1 in range(j, basesize[1]):
+                            ccol_indices[j1 + 1] += 1
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2550]: nit: We can probably simplify this with tensors + masking + advanced indexing.

File: torch/testing/_internal/common_utils.py (hunk 6 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2609]: Applied with some modifications.

File: torch/testing/_internal/common_utils.py (hunk 7 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2515]: generate_simple_inputs now uses tensors

File: torch/testing/_internal/common_utils.py (hunk 8 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2524]: generate_simple_inputs now uses tensors

File: torch/testing/_internal/common_utils.py (hunk 9 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2550]: generate_simple_inputs now uses tensors

File: test/test_sparse.py [hunk 10]
```diff
@@ -4049,6 +4049,58 @@ def test_basic(self):
         self.assertEqual(r.values(), torch.empty(0, 4, device='meta'))
 
 
+class TestSparseAny(TestCase):
+
+    def test_generate_simple_inputs(self):
+        # Temporarily disable BSC and BSC layouts as these don't support select yet
+        layouts = [torch.strided, torch.sparse_coo, torch.sparse_csr, torch.sparse_csc, torch.sparse_bsr, torch.sparse_bsc][:-2]
```
Comment (amjames) [inline test/test_sparse.py:4056]: Is the select requirement coming from `assertEqual`? 

Is there some workaround we could use to test the generator for those layouts? I don't feel great about them being available to pull from the generator without testing.

File: torch/testing/_internal/common_utils.py [hunk 11]
```diff
@@ -2448,6 +2448,285 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
...
+            r.add_(base)
+            return r
+
+        if patterns is None:
+            # A pattern is a 3-tuple with items:
+            # - a deep list of integers that define the sparsity
+            #   patterns of the generated inputs: zero values
+            #   correspond to unspecified elements/blocks and non-zero
+            #   values the specified elements while elements with the
+            #   same value belong to the same block in some cases.
```
Comment (amjames) [inline torch/testing/_internal/common_utils.py:2634]: > `same value belong to the same block in some cases.`

This is a bit confusing what are the conditions for "some cases" and what happens in the other cases (where specified values do not fall in the same block, or the layout is not blocked)?

**Review by amjames — COMMENTED** [review]
One minor comment inline. Since this is pretty complicated I think it is important that the comments are very clear.

File: torch/testing/_internal/common_utils.py (hunk 11 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py]: I have updated the comments.

File: test/test_sparse.py [hunk 12]
```diff
@@ -4049,6 +4049,58 @@ def test_basic(self):
...
+
+        generators = list(map(self.generate_simple_inputs, layouts))
+
+        tested_combinations = set()
+        while True:
+            try:
+                tensors = [next(g) for g in generators]
```
Comment (cpuhrsch) [inline test/test_sparse.py:4063]: Can't you use [`zip`](https://docs.python.org/3.3/library/functions.html#zip) for this? Or is that not the intend?

Comment (pearu) [reply test/test_sparse.py:4063]: Yes, using `zip` is a good idea.

Comment (pearu) [reply test/test_sparse.py]: Done. Thanks for the suggestion, @cpuhrsch !

**Review by cpuhrsch — APPROVED** [review]
